### PR TITLE
Automatically migrate users to multi-auth

### PR DESCRIPTION
**Pull request by Hamms**
Configure the User model to call `migrate_to_multi_auth` automatically upon user creation! Some notes:

- we are using `after_create_commit` instead of `after_create`, because modifying fields of the created instance in `after_create` doesn't work quite as expected (see [Underrated and enhanced dirty attributes changes in Rails 5.1](https://www.levups.com/en/blog/2017/undocumented-dirty-attributes-activerecord-changes-rails51.html) for more info)
- also add some more validation to authentication options; specifically, enabling anti-:panda_face:  email validation and making sure we validate the primary_contact_info even if it's not one of the authentication_options (a state which it shouldn't be possible to get into, but which a handful of our tests are testing for)
- adds the `demigrated` FactoryGirl trait for users. I have removed _most_ of the tests that are testing functionality specific to demigrated users, but there are still some scenarios (mostly takeover logic) for which it makes sense to continue to test. These tests (and the associated trait) can probably all be removed once we complete the migration of all existing user data.
- The vast, vast majority (98.4% by line count) of the changes in this PR are updates to existing tests. Some of these were testing the differences between migrated and unmigrated users, but many of them were simply trying to access or modify user-specific data from/in a place it no longer exists.

### CODE REVIEW

**Comment by islemaster** [comment]
PR description please! :grin:

File: dashboard/app/models/user.rb [hunk 1]
```diff
@@ -209,21 +209,26 @@ class User < ActiveRecord::Base
...
   # This custom validator makes email collision checks on the AuthenticationOption
   # model also show up as validation errors for the email field on the User
   # model.
   # There's probably some performance cost in additional queries here - once
   # we are fully migrated to multi-auth, we may want to remove this code and
   # check that we handle validation errors from AuthenticationOption everywhere.
   validate if: :migrated? do |user|
+    if user.primary_contact_info && !user.primary_contact_info.valid?
+      user.primary_contact_info.errors.each {|k, v| user.errors.add k, v}
+    end
```
Comment (islemaster) [inline dashboard/app/models/user.rb:221]: :tada: Yay surfacing validation errors!

File: dashboard/test/controllers/registrations_controller/demigrate_from_multi_auth_test.rb [hunk 2]
```diff
@@ -22,6 +22,7 @@ class DemigrateFromMultiAuthTest < ActionDispatch::IntegrationTest
 
   test "demigrate_from_multi_auth is a no-op for an unmigrated user" do
     teacher = create :teacher
+    teacher.demigrate_from_multi_auth
     refute teacher.migrated?
```
Comment (islemaster) [inline dashboard/test/controllers/registrations_controller/demigrate_from_multi_auth_test.rb:25]: Looks like this test no longer tests what it says it's testing.  Use `:demigrated` trait _and_ call `demigrate_from_multi_auth`?

File: dashboard/test/controllers/registrations_controller/migrate_to_multi_auth_test.rb [hunk 3]
```diff
@@ -9,6 +9,7 @@ class MigrateToMultiAuthTest < ActionDispatch::IntegrationTest
 
   test "migrate_to_multi_auth migrates an unmigrated user" do
     teacher = create :teacher
+    teacher.demigrate_from_multi_auth
```
Comment (islemaster) [inline dashboard/test/controllers/registrations_controller/migrate_to_multi_auth_test.rb]: Worth using `:demigrated` trait here for searchability later?

File: dashboard/test/controllers/registrations_controller/set_email_test.rb [hunk 4]
```diff
@@ -83,14 +86,25 @@ class SetEmailTest < ActionDispatch::IntegrationTest
     # passwords.  Examples of users without passwords are users that authenticate
     # via oauth (a third-party account), or students with a picture password.
 
-    test "editing email of student-without-password is not allowed" do
+    test "editing email of unmigrated student-without-password is not allowed" do
+      unmigrated_student_without_password = create :student
+      unmigrated_student_without_password.demigrate_from_multi_auth
```
Comment (islemaster) [inline dashboard/test/controllers/registrations_controller/set_email_test.rb]: Ditto `:demigrated` trait here?

File: dashboard/test/controllers/registrations_controller/set_user_type_test.rb [hunk 5]
```diff
@@ -23,9 +23,6 @@ class SetUserTypeTest < ActionDispatch::IntegrationTest
       assert_response :bad_request
     end
 
-    #
-    # Tests for non-migrated users
-    #
```
Comment (islemaster) [inline dashboard/test/controllers/registrations_controller/set_user_type_test.rb]: Did you opt to remove the non-migrated user tests here? I can't wait to do this, but it seems like we're not _quite_ ready to do so until we've migrated everyone.

File: dashboard/test/factories/factories.rb [hunk 6]
```diff
@@ -365,7 +341,7 @@
     end
 
     trait :with_migrated_google_authentication_option do
-      after(:create) do |user|
+      after(:create_commit) do |user|
```
Comment (islemaster) [inline dashboard/test/factories/factories.rb:344]: :scream: I remember this being kind of a mess, but never dug into the root cause.  Nice investigation on this!

File: dashboard/test/models/authentication_option_test.rb [hunk 7]
```diff
@@ -54,8 +55,8 @@ class AuthenticationOptionTest < ActiveSupport::TestCase
   test 'user can have multiple authentication options' do
     assert_creates(User) do
       user = create(:user, :with_google_authentication_option, :with_clever_authentication_option)
-      assert user.authentication_options.length == 2
-      assert user.authentication_options.first.hashed_email == user.hashed_email
+      assert_equal 3, user.authentication_options.length
+      assert_equal user.hashed_email, user.authentication_options.first.hashed_email
```
Comment (islemaster) [inline dashboard/test/models/authentication_option_test.rb:59]: Why do we expect the user to have 3 authentication options now instead of 2?

**Review by islemaster — COMMENTED** [review]
Amazing.  Thank you for working through the test suite and all of the single-auth assumptions we left behind, and my apologies for not warning about this in advance; @ewjordan knew this was going to be a difficult step and warned me about it, but I forgot that it was right where we left this work.

I've got some questions about places where some test expectations changed re: number of `AuthenticationOption`s, and why you've chosen to remove single-auth tests in some places and preserve them in others.

File: dashboard/test/models/authentication_option_test.rb (hunk 7 above)
Comment (Hamms) [reply dashboard/test/models/authentication_option_test.rb:59]: the two `with_authentication_option` traits each add an authentication option on top of the one that's created by migrating a standard user. In several places, I resolved this by creating the user with SSO properties directly (which would then be transformed into the appropriate authentication option), but in this case I would have had to do that with only one of the desired auth options. I decided to leave them both as "extras" for parallelism.

File: dashboard/test/controllers/registrations_controller/set_user_type_test.rb (hunk 5 above)
Comment (Hamms) [reply dashboard/test/controllers/registrations_controller/set_user_type_test.rb]: I did! Although looking at this again, I can see how these tests are definitely still valuable. I'll restore em, and just update them all to use `demigrated`

**Comment by Hamms** [comment]
PTAL!

**Review by islemaster — APPROVED** [review]
:tada: Let's do this!

File: dashboard/app/models/user.rb [hunk 8]
```diff
@@ -209,21 +209,26 @@ class User < ActiveRecord::Base
...
+
     user.authentication_options.each do |ao|
       unless ao.valid?
         ao.errors.each {|k, v| user.errors.add k, v}
       end
     end
   end
 
+  has_many :teacher_feedbacks, foreign_key: 'teacher_id', dependent: :destroy
```
Comment (maddiedierker) [inline dashboard/app/models/user.rb:230]: should this line be part of this PR? it seems unrelated, but maybe i'm just confused about its purpose

**Review by maddiedierker — APPROVED** [review]
one inline question, but otherwise this looks great! :shipit:

File: dashboard/app/models/user.rb (hunk 8 above)
Comment (Hamms) [reply dashboard/app/models/user.rb:230]: oh, this was just a minor reorganization. For some reason, this line was in between the authentication options relationship declaration and the authentication-option-specific validation. I just moved it so all the AO stuff is together.

Comment (maddiedierker) [reply dashboard/app/models/user.rb:230]: oh, that makes sense. thank you!

**Pull request merged**